neurons/miner.py

```diff
@@ -129,143 +137,238 @@ def get_config():
+
+                # Forward pass: compute the model output and loss
+                outputs = model(inputs, labels=inputs)
+
+                loss = outputs.loss / accumulation_steps  # Scale loss
+                loss.backward()  # Accumulate gradients
+
+                if (i + 1) % accumulation_steps == 0:
+                    n_acc_steps += 1
+                    optimizer.step()  # Perform a single optimization step
+                    optimizer.zero_grad()  # Clear gradients
+                    bt.logging.success(
+                        f"Step: {n_acc_steps} loss: {outputs.loss.detach().item()}"
                     )
+                    if use_wandb:
+                        wandb_run.log(
+                            {"loss": outputs.loss.detach(), "n_batches": n_batches},
+                            step=n_acc_steps,
+                        )
 
-            torch.cuda.empty_cache()
+                torch.cuda.empty_cache()
 
-            # Log the loss for the current step
-            n_batches += 1
-            global_step += 1
-            epoch_loss += outputs.loss.detach().item()
+                n_batches += 1
+                global_step += 1
+                epoch_loss += outputs.loss.detach().item()
 
-        # Calculate the average loss for the epoch
-        avg_loss = epoch_loss / n_batches
+            # Calculate the average loss for the epoch
+            avg_loss = epoch_loss / n_batches
 
-        # Log the average loss for the epoch
-        bt.logging.success(f"Epoch: {epoch_step} average loss: {avg_loss}")
-        epoch_step += 1
+            # Log the average loss for the epoch
+            bt.logging.success(f"Epoch: {epoch_step} average loss: {avg_loss}")
+            epoch_step += 1
 
-        # Check if the average loss of this epoch is the best we've seen so far
-        if avg_loss < best_avg_loss * (1 - pretrain.timestamp_epsilon):
-            best_avg_loss = avg_loss  # Update the best average loss
-            bt.logging.success(f"New best average loss: {best_avg_loss}.")
-            # Save the model to your mining dir.
+            # Check if the average loss of this epoch is the best we've seen so far
+            if avg_loss < best_avg_loss * (1 - pt.timestamp_epsilon):
+                best_avg_loss = avg_loss  # Update the best average loss
+
+                bt.logging.success(f"New best average loss: {best_avg_loss}.")
+
+                # Save the model to your mining dir.
+                bt.logging.success(f"Saving model to path: {model_path}.")
+                miner_actions.save(model, model_path)
+
+                # Also upload this better model to wandb, if enabled.
+                if use_wandb:
+                    wandb_run.save(glob_str=model_path)
+
+        bt.logging.success("Finished training")
+        # Push the model to your run.
+        if not config.offline:
+            if best_avg_loss < config.avg_loss_upload_threshold:
```

Comment:
Should we adjust this by the pt.timestamp_epsilon as well? Or is that implied in the config value?

Reply:
This is simply the threshold that the user specifies. I don't think we need to adjust.